pkg/controller/jobs/pod/pod_controller_test.go

```diff
@@ -3178,3 +3179,135 @@ func TestGetWorkloadNameForPod(t *testing.T) {
+			Label("kueue.x-k8s.io/managed", "true").
+			KueueFinalizer().
+			Group("test-group").
+			CreationTimestamp(now.Add(time.Minute)).
+			GroupTotalCount("2").
+			Obj(),
+		*basePodWrapper.
+			Clone().
+			Name("excessPod").
+			Label("kueue.x-k8s.io/managed", "true").
+			KueueFinalizer().
+			Group("test-group").
+			CreationTimestamp(now.Add(time.Minute * 2)).
+			GroupTotalCount("2").
+			Obj(),
+	}
+
+	wl := *utiltesting.MakeWorkload("test-group", "ns").
+		PodSets(
+			*utiltesting.MakePodSet("60bc72d3", 2).
+				Request(corev1.ResourceCPU, "1").
+				Obj(),
+		).
+		Queue("user-queue").
+		OwnerReference(corev1.SchemeGroupVersion.WithKind("Pod"), "pod", "test-uid").
+		OwnerReference(corev1.SchemeGroupVersion.WithKind("Pod"), "pod2", "test-uid").
+		OwnerReference(corev1.SchemeGroupVersion.WithKind("Pod"), "excessPod", "test-uid").
+		ReserveQuota(utiltesting.MakeAdmission("cq").AssignmentPodCount(1).Obj()).
+		Admitted(true).
+		Obj()
+
+	clientBuilder := utiltesting.NewClientBuilder()
+	if err := SetupIndexes(ctx, utiltesting.AsIndexer(clientBuilder)); err != nil {
+		t.Fatalf("Could not setup indexes: %v", err)
+	}
+
+	for i := range pods {
+		clientBuilder = clientBuilder.WithObjects(&pods[i])
+	}
+
+	kcBuilder := clientBuilder.
+		WithStatusSubresource(&wl).
+		WithInterceptorFuncs(interceptor.Funcs{
+			Update: func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.UpdateOption) error {
+				if reqcount == 0 {
+					reqcount += 1
+					return connRefusedEerrMock
+				}
+				reqcount = 0
+				return client.Update(ctx, obj, opts...)
+			},
+			Delete: func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.DeleteOption) error {
+				if reqcount == 0 {
+					reqcount += 1
+					return connRefusedEerrMock
+				}
+				reqcount = 0
+				return client.Delete(ctx, obj, opts...)
+			},
+		})
+
+	kClient := kcBuilder.Build()
+	if err := kClient.Create(ctx, &wl); err != nil {
+		t.Fatalf("Could not create workload: %v", err)
+	}
+
+	recorder := record.NewBroadcaster().NewRecorder(kClient.Scheme(), corev1.EventSource{Component: "test"})
+	reconciler := NewReconciler(kClient, recorder)
+
+	reconcilerImpl := reconciler.(*Reconciler)
+	excessPodExpectationsStore := reconcilerImpl.expectationsStore
+
+	reconcileRequest := reconcileRequestForPod(&pods[0])
+	podGroupKey := types.NamespacedName{
+		Name:      reconcileRequest.NamespacedName.Name,
+		Namespace: strings.Split(reconcileRequest.NamespacedName.Namespace, "/")[1],
+	}
+
+	// Reconcile for the first time. It'll try  to remove the finalizers but fail
+	_, err := reconciler.Reconcile(ctx, reconcileRequest)
+	if diff := cmp.Diff(connRefusedEerrMock, err, cmpopts.EquateErrors()); diff != "" {
+		t.Errorf("Expected reconcile error (-want,+got):\n%s", diff)
+	}
+
+	// The pod should've been observed even though the update failed
+	if satisfied := excessPodExpectationsStore.Satisfied(ctrl.LoggerFrom(ctx), podGroupKey); !satisfied {
```

Comment:
I think checking the expectations is not necessary actually. If they were not cleaned, the next Reconcile would not be able to make progress. 

OTOH these are unit tests, so getting into the weeds is ok. Still, I think we could start simple, will less code, yet covering the bug if the lines were missing.

Reply:
Yeah, you are right, the expectations check is not necessary since the controller is waiting to observe previous operations on pods 👍🏻